### PR TITLE
🚀 Order Approval | Requisitions & Approvals & Budget Widget

### PR DESCRIPTION
**Pull request by MaxKless**
## **:rocket: Order Approval | Requisitions & Approvals & Budget Widget**

IAD: https://ytqc9o.axshare.com/#id=d49m0o&p=requisitions&g=1
Budget Widget: #309

- [x] basic widget functionality
- [x] manage budgets only for admin
- [x] remove dashboard icons for b2b
- [x] replace isOrderApprovalEnabled with pipe
- [x] confirm message texts again
- [x] style widgets 
- [ ] re-evaluate widget data loading

## PR Type
Feature

## Does this PR Introduce a Breaking Change?

Not yet

## Other Information
*Notes*
* lazy component in `order-approval` extension or implemented in `order-approval` project but usable in PWA core?

### CODE REVIEW

**Comment by MaxKless** [comment]
@marschmidt89 could you provide german and french translations for the following localization keys?
account.quotes.widget.responded.label
account.requisitions.widget.approvals_title
account.requisitions.widget.budget_title
account.requisitions.widget.order_total
account.requisitions.widget.pending
account.requisitions.widget.requisitions_title
account.requisitions.widget.view_all_approvals
account.requisitions.widget.view_all_budgets
account.requisitions.widget.view_all_requisitions

Thanks!

**Comment by SGrueber** [comment]
Issue related to the ui:
- [ ] Approval widget should only be visible, if the user has approver permission
- [ ] responsive behavior of the widgets (size paddings etc.)
- [ ] if the order total is N/A (because to item number = 0) it should either be hidden or 0 - no N/A display

**Review by M-Behr — CHANGES_REQUESTED** [review]
- [ ] The white "View All..." links on the blue cards need a hover effect (bootstrap default 15% darker)
- [ ] My Budgets should be visible in case I have the buyer role
- [ ] Manage Budgets link should only be visible in case of admin role
- [ ] My Approvals should reflect the amount of items, which I see under Approvals > Waiting for your Approval (currently the amount is only correct for the person, how requested the order)
Here is an example:
Approval List
Related to the approval list, Conny should see 3 pending approvals and an order total of $30,019.12 in the my approvals widget. Currently she only sees 1 pending approval. I think the requisitions are displayed here in the approval widget instead of approvals. Because the link also leads to the false page as Silke mentioned.

File: src/app/shared/components/common/info-box/info-box.component.ts [hunk 1]
```diff
@@ -23,4 +23,9 @@ export class InfoBoxComponent {
    * Router link for Editing the displayed data. If a routerLink is given a link is displayed to route *to an edit page
    */
   @Input() editRouterLink: string;
+
+  /**
+   * Additional css classes to be passed to the infobox div
+   */
+  @Input() cssClass: string;
```
Comment (SGrueber) [inline src/app/shared/components/common/info-box/info-box.component.ts:30]: should be optional

File: src/styles/pages/checkout/checkout.scss [hunk 2]
```diff
@@ -94,6 +94,42 @@
...
+        font-size: 60px;
+        line-height: 1;
+      }
+      .row:last-of-type {
+        margin-top: 0.5rem;
+      }
+    }
+
+    &.infobox-widget {
+      padding-bottom: 15px;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:127]: $space-default

File: src/styles/pages/checkout/checkout.scss [hunk 3]
```diff
@@ -94,6 +94,42 @@
         padding-left: $space-default * 2;
       }
     }
+
+    &.infobox-color-widget {
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:98]: general: don't use fix font sizes, padding etc. but Bootstrap Variables instead

File: src/styles/pages/checkout/checkout.scss [hunk 4]
```diff
@@ -94,6 +94,42 @@
...
+      padding-bottom: 15px;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:99]: padding-bottom: $space-default

File: src/styles/pages/checkout/checkout.scss [hunk 5]
```diff
@@ -94,6 +94,42 @@
...
+      font-size: 14px;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:100]: unnecessary

File: src/styles/pages/checkout/checkout.scss [hunk 6]
```diff
@@ -94,6 +94,42 @@
...
+      color: $text-color-inverse;
+      text-transform: uppercase;
+      background: $CORPORATE-PRIMARY;
+      h3 {
+        font-size: 1.09375rem;
+        color: $text-color-inverse;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:106]: inherit

File: src/styles/pages/checkout/checkout.scss [hunk 7]
```diff
@@ -94,6 +94,42 @@
...
     }
+
+    &.infobox-color-widget {
+      padding-bottom: 15px;
+      font-size: 14px;
+      color: $text-color-inverse;
+      text-transform: uppercase;
+      background: $CORPORATE-PRIMARY;
+      h3 {
+        font-size: 1.09375rem;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:105]: unnessesssary

File: src/styles/pages/checkout/checkout.scss [hunk 8]
```diff
@@ -94,6 +94,42 @@
...
+        color: $text-color-inverse;
+      }
+      a {
+        color: $text-color-inverse;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:109]: inherit

File: src/styles/pages/checkout/checkout.scss [hunk 9]
```diff
@@ -94,6 +94,42 @@
...
+        text-decoration: underline;
+        text-transform: none;
+      }
+      .infobox-count {
+        font-size: 36px;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:114]: $font-size-lg * 2 or in rem

File: src/styles/pages/checkout/checkout.scss [hunk 10]
```diff
@@ -94,6 +94,42 @@
...
+        line-height: 1;
+      }
+      .infobox-count-large {
+        font-size: 60px;
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:118]: $font-size-lg * 3 or in rem

File: src/styles/pages/checkout/checkout.scss [hunk 11]
```diff
@@ -94,6 +94,42 @@
...
+      }
+      .row:last-of-type {
+        margin-top: 0.5rem;
+      }
+    }
+
+    &.infobox-widget {
+      padding-bottom: 15px;
+
+      .row:last-of-type {
```
Comment (SGrueber) [inline src/styles/pages/checkout/checkout.scss:129]: why is this necessary - i couldn`t recognize a difference

File: src/app/core/models/price/price.helper.ts [hunk 12]
```diff
@@ -58,4 +58,12 @@ export class PriceHelper {
       value: Math.round((p1.value + p2.value) * 100) / 100,
     };
   }
+
+  static empty(currency?: string): Price {
```
Comment (SGrueber) [inline src/app/core/models/price/price.helper.ts:62]: missing test

File: projects/organization-management/src/app/components/user-budget/user-budget.component.html [hunk 13]
```diff
@@ -1,21 +1,21 @@
...
+      <dd class="col-5 column-price font-weight-bold text-right">{{ 'account.budget.unlimited' | translate }}</dd>
     </ng-template>
   </dl>
 
   <ng-container *ngIf="budget.budget; else noBudget">
     <dl class="row dl-horizontal dl-separator">
-      <dt class="col-12 col-sm-7">{{ 'account.budget.type.' + budget.budgetPeriod + '.label' | translate }}</dt>
-      <dd class="col-5 column-price">{{ budget.budget | ishPrice }}</dd>
+      <dt class="col-7">{{ 'account.budget.type.' + budget.budgetPeriod + '.label' | translate }}</dt>
+      <dd class="col-5 column-price font-weight-bold text-right">{{ budget.budget | ishPrice }}</dd>
```
Comment (SGrueber) [inline projects/organization-management/src/app/components/user-budget/user-budget.component.html:15]: should not bold on user list - pls clarify with iad/vd if this is ok

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: projects/requisition-management/src/app/components/approval-widget/approval-widget.component.ts [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+  numPendingApprovals$: Observable<number>;
+  totalAmountApprovals$: Observable<Price>;
+
+  requisitionsLoading$: Observable<boolean>;
+
+  constructor(private requisitionFacade: RequisitionManagementFacade) {}
+
+  ngOnInit() {
+    const pendingApprovals$ = this.requisitionFacade.approverPendingRequisitions$().pipe(
+      filter(reqs => reqs.length > 0),
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply projects/requisition-management/src/app/components/approval-widget/approval-widget.component.ts]: shareReplay is removed, filter is needed so the sum and counting logic is not executed on empty observables

File: projects/requisition-management/src/app/store/requisitions/requisitions.effects.ts [hunk 15]
```diff
@@ -29,9 +29,9 @@ export class RequisitionsEffects {
     this.actions$.pipe(
       ofType(loadRequisitions),
       mapToPayload(),
-      switchMap(({ view, status }) =>
+      mergeMap(({ view, status }) =>
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply projects/requisition-management/src/app/store/requisitions/requisitions.effects.ts:32]: I used to need mergeMap, but now switchMap is correct, yes :)

File: src/styles/pages/checkout/checkout.scss (hunk 10 above)
Comment (MaxKless) [reply src/styles/pages/checkout/checkout.scss:118]: according to Inken it should be 60px so I'm doing 3.75rem

File: src/styles/pages/checkout/checkout.scss (hunk 11 above)
Comment (MaxKless) [reply src/styles/pages/checkout/checkout.scss:129]: This is necessary for more space between the budget widget content and link at the bottom

**Review by iwiederhold — APPROVED** [review]
THX Max! 
Very good collaboration!

**Review by M-Behr — CHANGES_REQUESTED** [review]
Well done! It works really good. I found two little things:
- [x] Page description needs more focus on B2B related features: "Everything you need to track your requisitions, quotes and orders, update your order templates, manage the budget and more is right here."
- [ ] My Budgets => My Budget

**Comment by MaxKless** [comment]
Hi @marschmidt89 could you provide a new translation for "account.overview.message_b2b.text", In addition, I changed "account.quotes.widget.submitted.label" to "Sent Requests", could you review the german and english translations? Thank you!

**Review by M-Behr — APPROVED** [review]

File: projects/requisition-management/src/app/models/requisition/requisition.model.ts [hunk 16]
```diff
@@ -7,7 +7,7 @@ import { User } from 'ish-core/models/user/user.model';
 
 export type RequisitionStatus = 'PENDING' | 'APPROVED' | 'REJECTED';
 
-export type RequisitionViewer = 'buyer' | 'approver';
+export type RequisitionViewer = 'buyer' | 'approver' | 'both';
```
Comment (SGrueber) [inline projects/requisition-management/src/app/models/requisition/requisition.model.ts:10]: 'all' or an optional view (no view for all) would be better, wouldn't be?

File: projects/requisition-management/src/app/services/requisitions/requisitions.service.ts [hunk 17]
```diff
@@ -63,6 +66,16 @@ export class RequisitionsService {
       .pipe(map(data => this.requisitionMapper.fromListData(data)));
   }
 
+  private getRequisitionsForBoth(status?: RequisitionStatus): Observable<Requisition[]> {
```
Comment (SGrueber) [inline projects/requisition-management/src/app/services/requisitions/requisitions.service.ts:69]: getAllRequisitions

File: projects/requisition-management/src/app/services/requisitions/requisitions.service.ts [hunk 18]
```diff
@@ -49,6 +49,9 @@ export class RequisitionsService {
    * @returns        Requisitions of the customer with their main attributes. To get all properties the getRequisition call is needed.
    */
   getRequisitions(view?: RequisitionViewer, status?: RequisitionStatus): Observable<Requisition[]> {
+    if (view === 'both') {
```
Comment (SGrueber) [inline projects/requisition-management/src/app/services/requisitions/requisitions.service.ts:52]: test?

**Review by SGrueber — CHANGES_REQUESTED** [review]

File: projects/requisition-management/src/app/models/requisition/requisition.model.ts (hunk 16 above)
Comment (MaxKless) [reply projects/requisition-management/src/app/models/requisition/requisition.model.ts:10]: since the 'both' type comes with some very specific behaviour, I would rather have it be a seperate option. I will change it to 'all'

**Review by SGrueber — APPROVED** [review]
